### ISSUE TITLE
Support to rename pickFile label in dropdowns using RegEx

### ISSUE THREAD

**Issue by arathburgos05** [rioj7/command-variable#73]
# 📄 Brief introduction

First of all, thanks for the extension it has a lot of interesting features.

I would like to rename the labels when for example `pickFIle` is used. Instead of seeing a filename or a path I would like to see a specific pattern created from the same data in the path. Also I would like to know if there is any way to search for empty folders specifically.

I read the documentation but I didn't found that I want to do.
The extension is working as expected, this is just for the presentation of the data.

# ⚙ Use case

## ❓ Question 1

Let's a consider a simple task to echo in console:
```
            {
                "label": "echo for testBaseTestSelectionTmp (improving presentation)",
                "type": "shell",
                "command": "echo",
                "args": [
                  "${input:testBaseTestSelectionTmp}"
                ],
                "problemMatcher": [],
                "hide": false
            }
```

and this generic folder structure:
```
└── my_tests
     ├── testType1
     │    └── LOG
     │        ├── testName1
     │        │   └── src
     │        │       └── test.c
     │        └── testName2
     │            └── src
     │                └── test.c
     └── testType2
          └── LOG
              ├── testName3
              │   └── src
              │       └── test.c
              └── testName4
                  └── src
                      └── test.c
```

my input file in tasks is set as:
```
            {
                "id": "testBaseTestSelectionTmp",
                "type": "command",
                "command": "extension.commandvariable.file.pickFile",
                "args": {
                    "description": "[my_tests] Select one test to open it",
                    "include": "**/test.c",
                    "fromWorkspace": "test_base"
                }
            },
```

The output for the specific case is:
- testType1/LOG/testName1/src/test.c ${w:my_test}
- testType1/LOG/testName2/src/test.c ${w:my_test}
- testType2/LOG/testName3/src/test.c ${w:my_test}
- testType2/LOG/testName4/src/test.c ${w:my_test}

What is working for practical purposes. However, I would like to convert the label to something like:
- testType1/testName1
- testType1/testName2
- testType2/testName3
- testType2/testName4

Using RegEx I could use:
`(.*)/LOG/(.*)/src/.*`

and this for the renamed label configuration:
`"labelRenamed": "$1/$2"`

Which would be the desired output. However, I didn't find any way to edit the labels in the dropdowns. Is there any or plans to implement something like this one?

## ❓ Question 2

Considering the same previous case, I would like to know if there is any way to list only folders not looking for a file.
For instance, if I would like to know what are the parent folders of LOG folders, something like...
`*/LOG`

I would expect to see:
- testType1/LOG
- testType1/LOG
- testType2/LOG
- testType2/LOG

This is just a dummy case, I have other uses in mind. I tried with **include** and **find** but it didn't work. The problem looking for a file is that this won't work if folder is empty. For the moment I have my workspace using file search and **transform** feature, but this could fail when folder is empty.

Thanks for your attention. Good job :)

**Comment by rioj7** [comment]
Moved question 2 to #74

**Comment by rioj7** [comment]
@arathburgos05  you can try v1.58.0

----
part of the `README` file

If your project contains file paths like:

`testType1/LOG/testName1/src/test.c`

and you only want to show the Type and the Name but return the full path use the following `input` element

```json
    {
      "id": "filePickCTests",
      "type": "command",
      "command": "extension.commandvariable.file.pickFile",
      "args": {
        "display": "transform",
        "description": " Select one test to open",
        "include": "**/test.c",
        "labelTransform": {
          "text": "${relativeFile}",
          "apply": [
            {
              "find": "\\\\",
              "replace": "/",
              "flags": "g"
            },
            {
              "find": "(.*)/LOG/(.*)/src/.*",
              "replace": "$1/$2"
            }
          ]
        }
      }
    }
```

If using the same file paths as the previous example but you want to show and return the test type folders that have a `LOG` subdirectory

```json
    {
      "id": "filePickCTests",
      "type": "command",
      "command": "extension.commandvariable.file.pickFile",
      "args": {
        "display": "transform",
        "ydisplay": "relativePath",
        "description": "[my_tests] Select one test to open it",
        "include": "my_tests/**/test.c",
        "labelTransform": "valueTransform",
        "valueTransform": {
          "text": "${relativeFile}",
          "apply": [
            {
              "find": "\\\\",
              "replace": "/",
              "flags": "g"
            },
            {
              "find": "(.*/LOG)/.*",
              "replace": "$1"
            }
          ]
        }
      }
    }
```

**Comment by arathburgos05** [comment]
It worked properly. Thanks @rioj7 

Just a simple question, is it expected that "${ws:NAME} disappears when labelTransform is used?

**Comment by rioj7** [comment]
@arathburgos05  use `"descriptionTransform"`

`${ws:NAME}` is part of the description of the QuickPickItem

you can use the icons: `$(folder)`

**Comment by arathburgos05** [comment]
> @arathburgos05  use `"descriptionTransform"`
> 
> `${ws:NAME}` is part of the description of the QuickPickItem
> 
> you can use the icons: `$(folder)`

Really nice job.

Thank you so much!! Great extension!

**Comment by rioj7** [comment]
@arathburgos05  need to correct.

> Is it expected that `${ws:NAME}` disappears when labelTransform is used?

`${ws:NAME}` is part of description when `"display": "relativePath"`

Yes is disappears if you use `"display": "transform"`, then the default description is nothing/empty text

**Comment by arathburgos05** [comment]
@rioj7

Hi there, is it possible to enable "predefined variables" or it implicates a a lot of work?

For example here:
```
    {
      "id": "filePickCTests",
      "type": "command",
      "command": "extension.commandvariable.file.pickFile",
      "args": {
        "display": "transform",
        "ydisplay": "relativePath",
        "description": "[my_tests] Select one test to open it",
        "include": "my_tests/**/test.c",
        "labelTransform": "valueTransform",
        "valueTransform": {
          "text": "${file}",  // <-------- Focus here..
          "apply": [
            {
              "find": "\\\\",
              "replace": "/",
              "flags": "g"
            },
            {
              "find": "${relativeFile}", // <-------- .. and here
              "replace": ""
            }
          ]
        }
      }
    }
```

That example will return the `${workspaceFolder:Matches}` directory of the matches (for multiworkspaces, it is specially useful when we have a multiworkspace with folders inside folders added to the workspace -useful to taking independent c_cpp configurations for intellisense-). There may be many other implementations. I think this can do more versatile the regex searches and replaces.

I realized that you have support for "predefined variables" in commands like `"text"` but not in commands like `"find"`. Is there any workaround for that?

Specifically, the problem that I'm getting is this one (not getting uniform results with `${relativeFile}`):
![image](https://github.com/rioj7/command-variable/assets/25379464/a5295481-d4d8-4e98-a847-ff360db76f62)

```
                    "labelTransform": {
                        "text": "${relativeFile}",
                        "apply": [
                            {
                            "find": "\\\\",
                            "replace": "/",
                            "flags": "g"
                            },
                            {
                            "find": "(.*)/IN/(.*)/source/.*$",
                            "replace": "$1/$2"
                            }
                        ]
                    }
```

The problem is caused because I have a subfolder inside another subfolder in the workspace. However, I cannot delete it because it is need to take the `compile_commands.json` for that specific sub-folder.
![image](https://github.com/rioj7/command-variable/assets/25379464/fdf13a38-5119-4a9e-b720-f4025e65e68b)

The best solution would be to get support for "predefined variables" for `"find"` and `"replace"` as you have with `"text"`. If this is not viable I can continue looking for other workarounds.

BTW, I really love your extension. It makes my work easier :)
![image](https://github.com/rioj7/command-variable/assets/25379464/85ba46b4-f3f0-4cfa-b2e6-f85c05c2e6ec)
![image](https://github.com/rioj7/command-variable/assets/25379464/a3c97173-be94-4636-a1d9-0c535174d725)

**PD:** Do you have any donate platform? I would like to invite you some coffees :) I really appreciate your work.

**Comment by rioj7** [comment]
@arathburgos05 

You don't tell me what the result of the `labelTransform` is, you only tell it is not the text `OK`.

If you start the transform with `${file}` and replace the `\` with `/` and you then do a find for `${file}` it will not match.
The `${file}` variables are expanded in the `text`, `replace` and `find` property.

If you look in the doc for [supported variables](https://marketplace.visualstudio.com/items?itemName=rioj7.command-variable#variables) you see a number of the predefined variables are also implemented (related to file and workspace).

Try 

```json
  "labelTransform": {
    "text": "${file}",
    "apply": [
      {
        "find": "${file}",
        "replace": "OK"
       }
    ]
  }
```

----

What is the purpose of the 2 images in the previous comment?

Are you showing that the result you see is `source/application/test.c` instead of `OK`

----

How do you get these colored icons in the list for **Select the task to run**?

**Comment by arathburgos05** [comment]
@rioj7 

I could solve the problem with another workaround using RegEx. I tried also with your code but it didn't work.
With the code of you last comment I would expect a `OK` always. But not the case, I only receive the `${file}` for the matches (variable is being solved properly but `"replace"` is not replacing anything. I think the problem is related to `\\` interpretation for RegEx). I confirmed that variables are being solved in `"find"` and `"replace"` properly, so my guess is that problem is related to slash interpretation when it tries to replace using RegEx.

As I mentioned I could solved it using another way, so I don't have any problem if we close this topic. Anyways, I'll answer all your comments.

> What is the purpose of the 2 images in the previous comment?
Are you showing that the result you see is source/application/test.c instead of OK

Exactly, I'm getting the `"text"` content and `"replace"` is doing anything (maybe because slashes conflict).

> How do you get these colored icons in the list for Select the task to run?
Basically I use `"icon"` modifier in this way:
```
        "tasks": [
            {
                "hide": false,
                "label": "[TESTING] Print var",
                "type": "shell",
                "command": "echo \"hello world!\"",
                "problemMatcher": [],
                "icon": {
                    "id": "archive",
                    "color": "terminal.ansiMagenta"
                }
            }
        ],
```

There are a lot of icons:
![image](https://github.com/rioj7/command-variable/assets/25379464/63ea53ba-cba8-4d2b-a6b9-8b97b8d27a26)

And we have these color options:
![image](https://github.com/rioj7/command-variable/assets/25379464/84bfaf0d-24c3-4634-9c6f-3b05c36b20fd)

At the end this is that I wanted to do:
![image](https://github.com/rioj7/command-variable/assets/25379464/db70bac4-8889-45cd-8c03-d0ce746d1e2c)

Thanks 👍

**Comment by arathburgos05** [comment]
@rioj7 

> I think I have to add a flag to the variable to process it as literal text in a regex find. There are quite a lot of characters to escape.
Yes it would be great. For the moment I have this workaround. First replace `"\\\\"` by `"/"`, then apply it. At then I use a RegEx to look for the expression that I want using that argument number:
![image](https://github.com/rioj7/command-variable/assets/25379464/89e4d266-61d2-4bce-a8f1-af77ce3accb5)

----

Yes, thanks for the website. Maybe you already knew but this command hides the folders grunt, gulp, npm, cmake, cppbuild, etc. In my case I don't need them because my commands are specifics.
```
    "settings": {
        "task.autoDetect": "off",
    },
```
From this:
![image](https://github.com/rioj7/command-variable/assets/25379464/c80fdc97-e1c7-44ab-8930-511bc6ac0dc6)

To this:
![image](https://github.com/rioj7/command-variable/assets/25379464/104835c8-0bb0-4b80-b178-01e2cf9e1134)

Regards.

**Comment by arathburgos05** [comment]
@rioj7 

Hello, I have a question out of topic. I'm starting with VS Code extension development. Do you know if exists any way to share a text between two extension easily?

My idea was to use your extension to get a path, then send it to my extension by a setting. In my workspace basically I have it:
```
{
	"folders": [
		{
			"path": "any_path"
		}
	],
    "settings": {
        "testToIndexIntellisense": "some_path"
	},
	"tasks": {
		"version": "2.0.0",
		"tasks": [
            {
                "hide": false,
                "label": "my attempt to send a variable value to my custom extension",
                "type": "process",
                "command": "${command:myCustomExtension.myCustomCmdArgs}",
                "args": [ "some_arg3" ],
                "problemMatcher": [],
            }
		],
		"inputs": [
            {
                "id": "myTest",
                "type": "command",
                "command": "myCustomExtension.myCustomCmdArgs",
                "args": [
                    "myArg2"
                ]
            }
		]
	},
}
```

In my extension basically I have a print for `myCustomExtension.myCustomCmdArgs`. With `"args": [ "some_arg3" ],` I'm getting  `${command:myCustomExtension.myCustomCmdArgs}` and `any_path` as received arguments in my extension.

This is the code basically:
```
vscode.commands.registerCommand('myCustomExtension.myCustomCmdArgs', (arg) => {
	vscode.window.showInformationMessage(`Received argument: ${arg}`);
});
```

And If I call like `"command": "${input:myTest}",` instead of `"command": "${command:myCustomExtension.myCustomCmdArgs}",` It works. However, `"myArg2"` is not solved by any variable or call to another input (I read that this is not allowed by VS Code).

Do you know if there is any way to send a string between two extensions easily? My main idea was something like...
```
{
	"folders": [
		{
			"path": "any_path"
		}
	],
    "settings": {
        "testToIndexIntellisense": "some_path"
	},
	"tasks": {
		"version": "2.0.0",
		"tasks": [
            {
                "hide": false,
                "label": "my attempt to send a variable value to my custom extension",
                "type": "process",
                "command": "${command:myCustomExtension.myCustomCmdArgs}",
                "args": [ "${input:call_to_your_script}" ],
                "problemMatcher": [],
            }
		],
		"inputs": [
            {
                "id": "myTest",
                "type": "command",
                "command": "myCustomExtension.myCustomCmdArgs",
                "args": [
                    "myArg2"
                ]
            },
            {
                "id": "call_to_your_script",
                "type": "command",
                "command": "extension.commandvariable.file.pickFile",
                "args": {
                    "description": "Select one test",
                    "include": "makefile",
                    "showDirs": false,
                    "valueTransform": {
                        "text": "${file}",
                        "apply": [
                            {
                                "find": "\\\\",
                                "replace": "/",
                                "flags": "g"
                            }
                        ]
                    }
                }
            }
		]
	},
}
```

And at the end I was expecting that `"${input:call_to_your_script}"` was solved in tasks but it didn't happen (args are not being sent to my extension from tasks but it does in input).

Any suggestion/idea is welcome. My purpose is to update settings dynamically. I tried with your option of "remember" but it didn't work for `"settings": {}` in my workspace.

Thanks for your attention.
Regards.

**Comment by rioj7** [comment]
@arathburgos05

`args` in a task are not used in any `${command:...}` variable that is in the `command` property of the task. they are the arguments for the command send to the terminal.

If you have multiple `${command:...}` variables in the `command` property of the task, do they get all the same arguments?

If you use `"command": "${command:myCustomExtension.myCustomCmdArgs}"` `myCustomExtension.myCustomCmdArgs` is being called with an `args` parameter set to `undefined`.

The fact that that you can use variables in the `input` entries of task/launch and keybinding that use `"command": "extension.commandvariable.*"` is because I have implemented that in the extension.

----

> I'm getting `${command:myCustomExtension.myCustomCmdArgs}` and `any_path` as received arguments in my extension.

I don't know how you can get `any_path` as an argument.

I know that if you use `${command:...}` variables in launch configs you get the launch config object as argument.

Maybe there is a similar construct if used in tasks.

----

Do you want to be able to configure the source of the argument or do you want to be able to always call `pickFile` for the argument?

If you always want to call `pickFile` you can do that from the `myCustomExtension` extension with `vscode.executeCommand()` and you supply the `args` as parameter to the call. You can execute the command `myCustomExtension.myCustomCmdArgs` from the command palette if you add it to the `contribute::commands` section of `package.json`. No need to use tasks.

If you want it configured you have to put the configuration as part of the command arguments. You have no idea what the source location is of the call made to `myCustomExtension.myCustomCmdArgs`. It could be from inside another extension.

```jsonc
{
  // ....
  "tasks": {
    "version": "2.0.0",
    "tasks": [
      // ....
    ],
    "inputs": [
      {
        "id": "myTest",
        "type": "command",
        "command": "myCustomExtension.myCustomCmdArgs",
        "args": [
          "myArg2",
          {
            "type": "command",
            "command": "extension.commandvariable.file.pickFile",
            "args": {
              "description": "Select one test",
              "include": "makefile",
              "showDirs": false,
              "valueTransform": {
                "text": "${file}",
                "apply": [
                  {
                    "find": "\\\\",
                    "replace": "/",
                    "flags": "g"
                  }
                ]
              }
            }
          }
        ]
      }
    ]
  }
}
```

Somewhere in a task use `$(input:myTest}`.

Now in the `myCustomExtension.myCustomCmdArgs` code you have to analyze the `args` parameter and if the array element is not a string but an object with the correct properties call the given command and supply it with the `args` from the `args` property of the object.

----

`remember` does not use the `settings.json` to store the stuff it remembers. It is an object in the Command Variable extension.

**Comment by arathburgos05** [comment]
@rioj7 

Thanks a lot! You solved my problem with your comment.

It's my first contact with the VS Code development environment (for extensions). Do you know of a page or repository to see more code related to extension development? I consider myself self-taught and I like to read documentation.

I know that the VS Code page is very easy to understand and well explained, but perhaps you know of another resource that could be useful to me with more advanced codes.

**Comment by rioj7** [comment]
@arathburgos05 read/browse the examples and the API ref page to see what is possible and how to use it. Also you can examine the code of other extensions that have a feature you want to know how it is implemented and use a variant in your own extension. Also the extensions written by VSC team members are very insightful of how to use the API.

**Comment by rioj7** [comment]
@arathburgos05 

Implemented filters in #76 and now the label should be `OK`

```json
  "labelTransform": {
    "text": "${file}",
    "apply": [
      {
        "find": "${file|regexEscape}",
        "replace": "OK"
       }
    ]
  }
```
